src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/AvoidConstArrays.cs

```diff
@@ -0,0 +1,94 @@
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.AvoidConstArraysDescription), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            // Analyzes an argument operation
+            context.RegisterOperationAction(operationContext =>
+            {
+                IArgumentOperation? argumentOperation;
+
+                if (operationContext.Operation is IArrayCreationOperation arrayCreationOperation) // For arrays passed as arguments
+                {
+                    argumentOperation = arrayCreationOperation.GetAncestor<IArgumentOperation>(OperationKind.Argument);
+                    if (argumentOperation is null)
+                    {
+                        return;
+                    }
+                }
+                else if (operationContext.Operation is IInvocationOperation invocationOperation) // For arrays passed in extension methods, like in LINQ
+                {
+                    if (invocationOperation.Descendants().Any(x => x is IArrayCreationOperation))
+                    {
+                        // This is an invocation that contains an array in it
+                        // This will get caught by the first case in another cycle
+                        return;
+                    }
+
+                    argumentOperation = invocationOperation.Arguments.FirstOrDefault();
+                    if (argumentOperation is null || argumentOperation.Children.First() is not IConversionOperation conversionOperation
+                        || conversionOperation.Operand is not IArrayCreationOperation)
+                    {
+                        return;
+                    }
+                    arrayCreationOperation = (IArrayCreationOperation)conversionOperation.Operand;
+                }
+                else
+                {
+                    return;
+                }
+
+                // Must be literal array
+                if (!arrayCreationOperation.Children.First(x => x is IArrayInitializerOperation).Children.All(x => x is ILiteralOperation))
```

Comment:
@sharwell Is it fine to "manually" look up for children of some type?

Reply:
I'll update this. I didn't look closely about the properties that were included, and I can clean this up and make it more performant.

Reply:
Updated to this:
https://github.com/dotnet/roslyn-analyzers/blob/9f0e5ac03e33c510231389e396d24f6ce812178a/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/AvoidConstArrays.cs#L78